src/Microsoft.Health.Checkpoints/Storage/StorageCheckpointClient.cs

```
@@ -0,0 +1,143 @@
+        {
+            EnsureArg.IsNotNull(checkpoint);
+            EnsureArg.IsNotNullOrWhiteSpace(checkpoint.Partition);
+            EnsureArg.IsNotNullOrWhiteSpace(checkpoint.Identifier);
+
+            var lastProcessedDateTime = EnsureArg.IsNotNullOrWhiteSpace(checkpoint.LastProcessedDateTime.DateTime.ToString("MM/dd/yyyy hh:mm:ss.fff tt"), nameof(checkpoint.LastProcessedDateTime));
+            var lastProcessedIdentifier = checkpoint.LastProcessedIdentifier;
+
+            var blobName = $"{checkpoint.Partition}/{checkpoint.Identifier}";
+            var blobClient = _storageClient.GetBlobClient(blobName);
+
+            var metadata = new Dictionary<string, string>()
+            {
+                { _lastProcessedDateTime,  lastProcessedDateTime },
+                { _lastProcessedIdentifier,  lastProcessedIdentifier },
+            };
+
+            try
+            {
+                var blobRequestOptions = new BlobRequestConditions();
+                blobRequestOptions.IfMatch = new ETag(checkpoint.ETag);
+                BlobInfo result = await blobClient.SetMetadataAsync(metadata, blobRequestOptions);
+                checkpoint.ETag = result.ETag.ToString();
+            }
+            catch (RequestFailedException ex) when (ex.Status == (int)HttpStatusCode.PreconditionFailed)
+            {
+                _logger.LogWarning("The checkpoint's ETag does not match ETag provided.");
+            }
+            catch (RequestFailedException ex) when ((ex.ErrorCode == BlobErrorCode.BlobNotFound) || (ex.ErrorCode == BlobErrorCode.ContainerNotFound))
+            {
+                using (var blobContent = new MemoryStream(Array.Empty<byte>()))
+                {
+                    BlobContentInfo result = await blobClient.UploadAsync(blobContent, metadata: metadata).ConfigureAwait(false);
+                    checkpoint.ETag = result.ETag.ToString();
+                }
+            }
+            catch (Exception ex)
```

We would generally not swallow the generic exception.